Split the part that read the blockundo data from disk so the actual u…

deadalnix opened the pull request:
The first read the undo data from disk and the second does the undo. This ensure we don't depend on the state of the filesystem to undo and makes is easier to unit test. The patch also includes a basic unit test for undo.

ftrader commented on src/validation.cpp:
```diff
@@ -1506,61 +1506,68 @@ bool DisconnectBlock(const CBlock &block, CValidationState &state,
-            }
+        // Restore inputs.
+        if (i < 1) {
+            // Skipe the coinbase.
```
spelling error, should be 'Skip'

ftrader commented on src/validation.cpp:
```diff
@@ -1506,61 +1506,68 @@ bool DisconnectBlock(const CBlock &block, CValidationState &state,
 
-    // undo transactions in reverse order
+    // Undo transactions in reverse order.
     for (int i = block.vtx.size() - 1; i >= 0; i--) {
```
I would consider adding a sanity check that block.vtx.size() > 0

deadalnix replied:
It must because of the previous check (and no, blockUndo.vtxundo.size() cannot be size_t.max, there wouldn't be enough addressable memory for this to happen).

deadalnix replied:
That being said I think I can rewrite this in a way that do have this edge case.

ftrader commented on src/validation.cpp:
```diff
@@ -1506,61 +1506,68 @@ bool DisconnectBlock(const CBlock &block, CValidationState &state,
 
     if (pfClean) *pfClean = false;
```
can remove as pfClean is passed through to ApplyBlockUndo() and is initialized there if needed

deadalnix replied:
If we don't reach ApplyBlockUndo, we still need to set the flag as expected by the caller.

ftrader replied:
good point, I didn't consider that.

deadalnix commented:
Ok I added the copy right blurb, rewrote the loop so it doesn't have edge case when the size is 0 and fixed typo. I think we should keep setting pfClean as this  matter is case there is an error before we reach ApplyBlockUndo.

ftrader commented:
Please fix undo.h copyright year:

`// Copyright (c) 20017 The Bitcoin developers`

deadalnix commented:
Done.

ftrader commented:
All good from my side.
